Feature add additional web subscription data sources

bmello4688 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

#### Description
Instead of using an enum in SubscriptionDataSource to create the stream reader in the engine, change Subscription data source to an abstract class. The new abstract class will take a Func that returns a stream reader that will be called in the engine to match the current behavior of stream reader initialization in the engine. Derived classes will be created to create cleaner constructors. Derived classes are LocalFileSubscriptionDataSource, RemoteFileSubscriptionDataSource, WebSubscriptionDataSource, and RestSubscriptionDataSource. RestSubscriptionDataSource is derived from WebSubscriptionDataSource. WebSubscriptionDataSource should be used for any additional web protocols a person would want to use.

#### Related Issue
Closes #5697

#### Motivation and Context
In my case, I wanted to create a custom data feed that uses a gRPC stream and I couldn't do it without modifying the lean engine. I feel I was able to abstract out the stream reading pretty nicely. 

#### Requires Documentation Change
Need to update documentation around creating custom data feeds, universes, etc. Anything that uses the basedata class to return a SubscriptionDataSource.

#### How Has This Been Tested?
Unit tests and running the lean engine locally.

#### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Refactor (non-breaking change which improves implementation)
- [x] New feature (non-breaking change which adds functionality)

#### Checklist:
<!--- The following is a checklist of items that MUST be completed before a PR is accepted -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. <!--- If not applicable, please explain why --> Existing tests should suffice
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

<!--- Template inspired by https://www.talater.com/open-source-templates/#/page/99 -->

Martin-Molinero commented on Brokerages/QuantConnect.Brokerages.csproj:
```diff
@@ -40,7 +40,7 @@
     <PackageReference Include="NodaTime" Version="3.0.5" />
     <PackageReference Include="QuantConnect.IBAutomater" Version="2.0.39" />
-    <PackageReference Include="RestSharp" Version="106.6.10" />
+    <PackageReference Include="RestSharp" Version="106.11.7" />
```
We should avoid updating dependency unless required or if it's the purpose of the PR itself

Martin-Molinero commented on Common/Data/RestSubscriptionDataSource.cs:
```diff
@@ -0,0 +1,20 @@
+using QuantConnect.Data;
```
nit - license and documentation missing in new files

Martin-Molinero reviewed (commented):
Interesting idea! It seems like a breaking change for any implemented custom data outside of Lean though, it might be possible to make it backwards compatible. Another option could be directly adding a new `SubscriptionTransportMedium` and it's `StreamReader` implementation

bmello4688 commented:
What do you think now? I added backwards compatibility with Obsolete attribute tags and log errors so people will get notified they should migrate over to the new way. I did not want to keep adding on to the SubscriptionDataSource object because it will end up becoming overly bloated with anything new that needs to be added to it.

Martin-Molinero commented:
Hey @bmello4688!
Sorry for the delay coming back to you
> I did not want to keep adding on to the SubscriptionDataSource object because it will end up becoming overly bloated with anything new that needs to be added to it.

These are core Lean engine classes really, we only have a hand full which cover most use cases we know off, and don't really change.
If you think there is some missing let's please go ahead and add it as a new `SubscriptionTransportMedium` with it's collection of tests and docs so it can be used by all too, which would be awesome! 💯
I'll close this PR for now, since it's fallen outdated since we've moved out the implementations of the custom data sources

bmello4688 commented:
Hi Martin,

Yea I understand. It does cut through everything and I know it would be a pain to change to. That’s why I added the backwards compatibility so people could still use the old way or the new way. In my opinion, web protocols change a lot, which is why I didn’t want to just add a Grpc to the SubscriptionTransportMedium, but I guess in the future, it could always be addressed. I’ll add a new PR with gRPC support.

Brian

From: ***@***.***>
Sent: Tuesday, September 28, 2021 1:58 PM
To: ***@***.***>
Cc: ***@***.***>; ***@***.***>
Subject: Re: [QuantConnect/Lean] Feature add additional web subscription data sources (#5698)


Hey @bmello4688<https://github.com/bmello4688>!
Sorry for the delay coming back to you

I did not want to keep adding on to the SubscriptionDataSource object because it will end up becoming overly bloated with anything new that needs to be added to it.

These are core Lean engine classes really, we only have a hand full which cover most use cases we know off, and don't really change.
If you think there is some missing let's please go ahead and add it as a new SubscriptionTransportMedium with it's collection of tests and docs so it can be used by all too, which would be awesome! 💯
I'll close this PR for now, since it's fallen outdated since we've moved out the implementations of the custom data sources

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/QuantConnect/Lean/pull/5698#issuecomment-929493815>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AB3WTQUC66FS3SQWG75CM3LUEH647ANCNFSM47GISHNQ>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

Martin-Molinero commented:
> Hi Martin,
> 
> Yea I understand. It does cut through everything and I know it would be a pain to change to. That’s why I added the backwards compatibility so people could still use the old way or the new way. In my opinion, web protocols change a lot, which is why I didn’t want to just add a Grpc to the SubscriptionTransportMedium, but I guess in the future, it could always be addressed. I’ll add a new PR with gRPC support.
> 
> Brian

Ty Brian, sounds great.
